clerk_test.go

```diff
@@ -0,0 +1,344 @@
+}`,
+		resp.HTTPStatusCode,
+		resp.TraceID,
+		resp.Errors[0].Code,
+		resp.Errors[0].Message,
+		resp.Errors[0].LongMessage,
+	)
+	assert.JSONEq(t, expected, resp.Error())
+}
+
+// This is how you define a Clerk API resource which is ready to be
+// used by the library.
+type testResource struct {
+	APIResource
+	ID     string `json:"id"`
+	Object string `json:"object"`
+}
+
+// This is how you define types which can be used as Clerk API
+// request parameters.
+type testResourceParams struct {
+	APIParams
+	Name string `json:"name"`
+}
+
+// This is how you define a Clerk API resource which can be used in
+// API operations that read a list of resources.
+type testResourceList struct {
+	APIResource
+	Resources  []testResource `json:"data"`
+	TotalCount int64          `json:"total_count"`
+}
+
+// This is how you define a type which can be used as parameters
+// to a Clerk API operation that lists resources.
+type testResourceListParams struct {
+	APIParams
+	ListParams
+	Name string `json:"name"`
+}
+
+// We need to implement the Queryable interface.
+func (params testResourceListParams) Add(q url.Values) {
+	q.Set("name", params.Name)
+	params.ListParams.Add(q)
+}
+
+func TestBackendCall_RequestHeaders(t *testing.T) {
+	ctx := context.Background()
+	method := http.MethodPost
+	path := "/v1/resources"
+	secretKey := "sk_test_123"
+
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		require.Equal(t, method, r.Method)
+		require.Equal(t, path, r.URL.Path)
+
+		// The client sets the Authorization header correctly.
+		assert.Equal(t, fmt.Sprintf("Bearer %s", secretKey), r.Header.Get("Authorization"))
+		// The client sets the User-Agent header.
+		assert.Equal(t, "Clerk/v1 SDK-Go/v2.0.0", r.Header.Get("User-Agent"))
+		assert.Equal(t, "application/json", r.Header.Get("Content-Type"))
+		// The client includes a custom header with the SDK version.
+		assert.Equal(t, "go/v2.0.0", r.Header.Get("X-Clerk-SDK"))
+
+		_, err := w.Write([]byte(`{}`))
+		require.NoError(t, err)
+	}))
+	defer ts.Close()
+
+	// Set up a mock backend which triggers requests to our test server above.
+	SetBackend(NewBackend(&BackendConfig{
+		HTTPClient: ts.Client(),
+		URL:        &ts.URL,
+	}))
+
+	// Simulate usage for an API operation on a testResource.
+	// We need to initialize a request and use the Backend to send it.
+	SecretKey = secretKey
+	req := NewAPIRequest(method, path)
+	err := GetBackend().Call(ctx, req, &testResource{})
+	require.NoError(t, err)
+}
+
+// TestBackendCall_SuccessfulResponse_PostRequest tests that for POST
+// requests (or other mutating operations) we serialize all parameters
+// in the request body.
+func TestBackendCall_SuccessfulResponse_PostRequest(t *testing.T) {
+	ctx := context.Background()
+	name := "the-name"
+	rawJSON := `{"id":"res_123","object":"resource"}`
+
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
❓ Would a package like [httpmock](https://github.com/jarcoal/httpmock) help us in some way?

Reply:
Thanks for the suggestion, but I don't see any real benefit to adding this dependency. 

Unless you have a specific scenario in mind, or you feel that the tests are hard to follow.

To give you some context, I wasn't even sure if the dependency to `testify` was worth it. :smile:

Reply:
It's a perfectly valid testing approach, although I think it's more usually followed in cases where handlers need to be tested. I just noted this in case we think we can avoid some boilerplate, that's all.